sgx-types/src/ssa.rs

```diff
@@ -192,19 +193,80 @@ bitflags! {
         /// Protection Key induced fault.
         const PK = 1 << 5;
         /// EPCM induced fault.
-        const SGX = 1 << 5;
+        const SGX = 1 << 15;
     }
 }
 
+bitflags! {
+    /// Flags for a general protection fault;
+    /// The General Protection Fault sets an error code, which is the segment selector index
+    /// when the exception is segment related. Otherwise, 0.
+    /// For more, refer to: https://wiki.osdev.org/Exceptions
+    pub struct GenProtFault: u32 {
+        ///  Exception originated externally to the processor.
+        const E = 1 << 0;
+        /// Bits 1 and 2 together indicate whether the selector (bits 3-15) references a
+        /// descriptor in the GDT, IDT, or LDT.
+        const TBL1 = 1 << 1;
+        /// Bits 1 and 2 together indicate whether the selector (bits 3-15) references a
+        /// descriptor in the GDT, IDT, or LDT.
+        const TBL2 = 1 << 2;
+        /// Selector index (bits 3-15) for the GDT, IDT, or LDT.
+        const SEL3 = 1 << 3;
+        /// Selector index (bits 3-15) for the GDT, IDT, or LDT.
+        const SEL4 = 1 << 4;
+        /// Selector index (bits 3-15) for the GDT, IDT, or LDT.
+        const SEL5 = 1 << 5;
+        /// Selector index (bits 3-15) for the GDT, IDT, or LDT.
+        const SEL6 = 1 << 6;
+        /// Selector index (bits 3-15) for the GDT, IDT, or LDT.
+        const SEL7 = 1 << 7;
+        /// Selector index (bits 3-15) for the GDT, IDT, or LDT.
+        const SEL8 = 1 << 8;
+        /// Selector index (bits 3-15) for the GDT, IDT, or LDT.
+        const SEL9 = 1 << 9;
+        /// Selector index (bits 3-15) for the GDT, IDT, or LDT.
+        const SEL10 = 1 << 10;
+        /// Selector index (bits 3-15) for the GDT, IDT, or LDT.
+        const SEL11 = 1 << 11;
+        /// Selector index (bits 3-15) for the GDT, IDT, or LDT.
+        const SEL12 = 1 << 12;
+        /// Selector index (bits 3-15) for the GDT, IDT, or LDT.
+        const SEL13 = 1 << 13;
+        /// Selector index (bits 3-15) for the GDT, IDT, or LDT.
+        const SEL14 = 1 << 14;
+        /// Selector index (bits 3-15) for the GDT, IDT, or LDT.
+        const SEL15 = 1 << 15;
+    }
+}
+
+/// The `errcd` in ExceptionInfo could be a page fault or a general protection fault.
+/// This type represents either.
+#[derive(Debug)]
+pub enum Fault {
+    /// General Protection Fault with flags.
+    GP {
+        /// The flags provide extra information on the fault.
+        flags: GenProtFault,
+    },
+    /// Page Fault with flags and memory address.
+    PF {
+        /// Contains the page fault linear address that caused a page fault.
+        addr: u64,
+        /// The flags provide extra information on the fault.
+        flags: PageFault,
+    },
```

Comment:
Let's make these enum variants tuple structs. I think named fields is probably just adding clutter.